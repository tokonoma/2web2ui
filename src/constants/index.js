/* eslint-disable max-lines */
import { tokens } from '@sparkpost/design-tokens-hibana';

export const DEFAULT_REDIRECT_ROUTE = '/landing';
export const AFTER_JOIN_REDIRECT_ROUTE = '/onboarding/plan';
export const SIGN_UP_ROUTE = '/join';
export const AUTH_ROUTE = '/auth';
export const TFA_ROUTE = '/auth/tfa';
export const ENABLE_TFA_AUTH_ROUTE = '/auth/enable-tfa';
export const SSO_AUTH_ROUTE = '/auth/sso';
export const DASHBOARD_ROUTE = '/dashboard';
export const ONBOARDING_SENDINGDOMAIN_ROUTE = '/onboarding/sending-domain';
export const COOKIE_DOMAIN = '.sparkpost.com';

export const FORMATS = {
  DATE: 'MMM Do',
  LONG_DATE: 'MMM Do YYYY',
  SHORT_DATE: 'YYYY-MM-DD',

  TIME: 'h:mma',
  MILITARY_TIME: 'H:mm',
  ANTE_MILITARY_TIME: 'H:mma', // for 24 hour accidentally including am/pm (FE-61)

  DATETIME: 'MMM Do h:mma',
  LONG_DATETIME: 'MMM Do YYYY h:mma',
  LONG_DATETIME_ALT: 'MMM DD, YYYY [at] h:mma',

  INPUT_DATES: ['YYYY-MM-DD'],
  INPUT_TIMES: ['h:mma', 'H:mm', 'H:mma'],
  INPUT_DATETIMES: ['YYYY-MM-DD h:mma', 'YYYY-MM-DD H:mm', 'YYYY-MM-DD H:mma'],
};

export const RELATIVE_DATE_OPTIONS = ['hour', 'day', '7days', '10days', 'custom'];

export const LINKS = {
  SPC_EU_URL: 'https://app.eu.sparkpost.com',
  SPC_US_URL: 'https://app.sparkpost.com',
  SP_HOME_PAGE: 'https://www.sparkpost.com',
  API_DOCS: 'https://developers.sparkpost.com/api',
  DOCS: 'https://www.sparkpost.com/docs',
  PREMIUM_SUPPORT: 'https://www.sparkpost.com/contact-premium',
  ENTERPRISE_SUPPORT: 'https://www.sparkpost.com/contact-enterprise',
  RECIP_API: 'https://developers.sparkpost.com/api/recipient-lists.html',
  AB_TESTING_API: 'https://developers.sparkpost.com/api/ab-testing.html',
  SUBACCOUNTS: 'https://sparkpost.com/docs/user-guide/subaccounts/',
  SUBACCOUNTS_API: 'https://developers.sparkpost.com/api/subaccounts.html',
  TOU: 'https://www.sparkpost.com/policies/tou',
  DOMAIN_VERIFICATION:
    'https://www.sparkpost.com/docs/tech-resources/enabling-multiple-custom-tracking-domains',
  SENDING_REQS: 'https://www.sparkpost.com/docs/getting-started/requirements-for-sending-domains',
  SENDING_SETUP: 'https://www.sparkpost.com/docs/getting-started/setting-up-domains/',
  TFA_BACKUP_CODES: 'https://support.sparkpost.com/customer/portal/articles/1948449',
  IP_WARM_UP: 'https://support.sparkpost.com/customer/portal/articles/1972209-ip-warm-up-overview',
  DAILY_USAGE: 'https://www.sparkpost.com/docs/getting-started/what-counts-daily-monthly-usage',
  LEARN_MORE_TFA:
    'https://www.sparkpost.com/docs/my-account-and-profile/enabling-two-factor-authentication',
  ONBOARDING_SENDING:
    'https://www.sparkpost.com/docs/getting-started/getting-started-sparkpost/#preparing-your-from-address',
  RECAPTCHA_LIB_URL: 'https://www.google.com/recaptcha/api.js',
  DAILY_MONTHLY_QUOTA_LIMIT_DOC: 'https://support.sparkpost.com/customer/portal/articles/2030894',
  GETTING_STARTED_GUIDE: 'https://www.sparkpost.com/docs/getting-started/getting-started-sparkpost',
  ADMIN_BOUNCE: 'https://support.sparkpost.com/', // TODO Add this to the bounce page
  SSO_GUIDE: 'https://www.sparkpost.com/docs/my-account-and-profile/sso',
  ALERTS_DOCS: 'https://developers.sparkpost.com/api/alerts/#alerts-post-create-an-alert',
  ALERTS_SURVEY: 'https://goo.gl/forms/rnIuTvdVF2xhpKCy2',
  SNIPPET_API_DOCS: 'https://developers.sparkpost.com/api/snippets/',
  MANDATORY_TFA:
    'https://www.sparkpost.com/docs/my-account-and-profile/enabling-two-factor-authentication/',
  RECIPIENT_VALIDATION_ACCESS: 'https://www.sparkpost.com/rv-access',
  AUTO_IP_WARMUP_SETUP: 'https://www.sparkpost.com/docs/user-guide/automated-ip-warmup/',
  MAILBOX_PROVIDERS: 'https://www.sparkpost.com/docs/user-guide/alerts/#mailbox-providers',
};

export const ENTERPRISE_PLAN_CODES = ['ent1'];

export const AWS_COOKIE_NAME = 'aws-mkt';

export const DAILY_LIMIT_REQUEST_TEMPLATE = 'daily-limit-increase';

export const SPC_EU_TENANT = 'spceu';
export const SPC_TENANT = 'spc';

export const CROSS_LINK_MAP = {
  [SPC_TENANT]: { label: '', url: LINKS.SPC_US_URL },
  [SPC_EU_TENANT]: { label: 'EU', url: LINKS.SPC_EU_URL },
};

export const TYPEAHEAD_LIMIT = 100;
export const METRICS_API_LIMIT = 1000;

export const FORMS = {
  LOGIN: 'loginForm',
  SSO: 'ssoLoginForm',
  JOIN: 'joinForm',
  JOIN_PLAN: 'joinPlanForm',
  JOIN_SENDING_DOMAIN: 'joinSendingDomainForm',
  INVITE_USER: 'userInviteForm',
  EDIT_USER: 'userEditForm',
  EVENTS_SEARCH: 'eventsSearchForm',
  RV_ADDPAYMENTFORM: 'rvAddPaymentForm',
  RV_ADDPAYMENTFORM_UPLOADLISTPAGE: 'rvAddPaymentFormUploadListPage',
};

//These filters are the searchable filters in the 'more filters' modal on the Events Search page
export const EVENTS_SEARCH_FILTERS = {
  recipient_domains: {
    placeholder: 'list of domain parts (e.g., gmail, yahoo.co, yahoo.co.jp, .fr)',
    label: 'Recipient Domains',
  },
  from_addresses: {
    placeholder: 'list of full from addresses (e.g., john@mycompany.com)',
    label: 'From Addresses',
  },
  sending_domains: {
    placeholder: 'list of domain parts (e.g., joespizza, marketing.joespizza)',
    label: 'Sending Domains',
  },
  subjects: {
    placeholder: 'list of full words (e.g. happy birthday, christmas)',
    label: 'Subjects',
  },
  bounce_classes: {
    placeholder: 'list of bounce classification codes (e.g. 10, 30)',
    label: 'Bounce Classes',
  },
  reasons: {
    placeholder: 'list of full words in raw reasons (e.g. 5.4.7, IP, Mail rejected)',
    label: 'Reasons',
  },
  campaigns: {
    placeholder: 'list of full words in campaign ids (e.g. newsletter, fire-sale)',
    label: 'Campaign IDs',
  },
  templates: {
    placeholder: 'list of full words in template ids (e.g. newsletter, fire-sale)',
    label: 'Template IDs',
  },
  sending_ips: { placeholder: 'list of ip addresses (e.g., 10.0.0.4.25)', label: 'Sending IPs' },
  ip_pools: {
    placeholder: 'list of full words in ip pool id (e.g billing, administrative)',
    label: 'IP Pools',
  },
  subaccounts: { placeholder: 'list of subaccount ids (e.g. 346, 823)', label: 'Subaccount IDs' },
  messages: {
    placeholder: 'list of message ids (e.g. 00129004225c33cc4c45)',
    label: 'Message IDs',
  },
  transmissions: {
    placeholder: 'list of transmission ids (e.g. 337221619334066426)',
    label: 'Transmissions IDs',
  },
  ab_tests: {
    placeholder: 'list of full words in ab test ids (e.g. newsletter, fire-sale)',
    label: 'AB Test IDs',
  },
  ab_test_versions: {
    placeholder: 'list of version numbers; requires AB Test ID filter(e.g. 4)',
    label: 'AB Test Versions',
  },
};

export const ALL_EVENTS_FILTERS = {
  recipients: { placeholder: 'Filter by recipient email address', label: 'Recipients' },
  events: { label: 'Event' },
  ...EVENTS_SEARCH_FILTERS,
};

export const DEFAULT_PER_PAGE_BUTTONS = [10, 25, 50, 100];

export const ANALYTICS_CREATE_ACCOUNT = 'create account';
export const ANALYTICS_ADDON_IP = 'dedicated_ips';
export const ANALYTICS_PREMIUM_SUPPORT = 'premium-support';
export const ANALYTICS_ENTERPRISE_SUPPORT = 'enterprise-support';
export const ANALYTICS_ONBOARDING = 'onboarding';
export const ANALYTICS_ONBOARDING_CREATE_DOMAIN = 'create-sending-domain-verify-';
export const ANALYTICS_WHITELISTED_FORMS = [
  FORMS.LOGIN,
  FORMS.JOIN,
  FORMS.JOIN_PLAN,
  FORMS.JOIN_SENDING_DOMAIN,
];

export const MAILBOX_PROVIDERS = {
  apple: 'Apple',
  gmail: 'Gmail',
  google_apps: 'Google Apps',
  microsoft: 'Microsoft',
  verizon_media: 'Verizon Media',
  yahoo_uk: 'Yahoo UK',
  us_broadband: 'US Broadband',
  uk_providers: 'UK Providers',
  italian_providers: 'Italian Providers',
  russian_providers: 'Russian Providers',
  french_providers: 'French Providers',
  chinese_providers: 'Chinese Providers',
  german_providers: 'German Providers',
  canadian_providers: 'Canadian Providers',
  japanese_providers: 'Japanese Providers',
  other_european_providers: 'Other European Providers',
  other: 'Other Providers',
};

export const PLAN_TIERS = {
  test: 'Test Account',
  starter: 'Starter',
  premier: 'Premier',
};

export const RECIPIENT_VALIDATION_TIERS = [
  { volumeMin: 0, volumeMax: 5000, cost: 0.01, displayedCost: '$0.010' },
  { volumeMin: 5000, volumeMax: 10000, cost: 0.008, displayedCost: '$0.008' },
  { volumeMin: 10000, volumeMax: 50000, cost: 0.006, displayedCost: '$0.006' },
  { volumeMin: 50000, volumeMax: 100000, cost: 0.004, displayedCost: '$0.004' },
  { volumeMin: 100000, volumeMax: 250000, cost: 0.003, displayedCost: '$0.003' },
  { volumeMin: 250000, volumeMax: 750000, cost: 0.0015, displayedCost: '$0.0015' },
  { volumeMin: 750000, volumeMax: 1000000, cost: 0.001, displayedCost: '$0.0010' },
  { volumeMin: 1000000, volumeMax: Infinity, cost: 0.00075, displayedCost: '$0.00075' },
];

export const HIBANA_METRICS_COLORS = [
  tokens.color_red_700,
  tokens.color_purple_800,
  tokens.color_yellow_500,
  tokens.color_blue_700,
  tokens.color_teal_600,
  tokens.color_magenta_800,
  tokens.color_blue_900,
  tokens.color_green_800,
  tokens.color_red_600,
  tokens.color_purple_600,
  tokens.color_yellow_700,
  tokens.color_teal_700,
  tokens.color_magenta_700,
  tokens.color_green_900,
];

export { ROLES, SUBACCOUNT_ROLES, ROLE_LABELS } from './users';
export { CAMPAIGN_IDS } from './vwo';
