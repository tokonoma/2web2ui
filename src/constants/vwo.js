export const CAMPAIGN_IDS = {
  ONBOARDING_SENDINGDOMAIN: [2, 19],
  CREATE_SENDINGDOMAIN: [3, 19],
  VERIFY_SENDINGDOMAIN: [203],
};
