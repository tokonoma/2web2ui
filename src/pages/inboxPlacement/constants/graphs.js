import { tokens } from '@sparkpost/design-tokens-hibana';

// TODO: Once OG theme is removed, remove theme-specific objects containing color configuration
export const OG_PERCENT_WHEEL = {
  color: '#0CBAC7',
  background: '#EEEEEE',
};

export const HIBANA_PERCENT_WHEEL = {
  color: tokens.color_blue_700,
  background: tokens.color_gray_300,
};
