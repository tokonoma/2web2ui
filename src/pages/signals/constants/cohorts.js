import { tokens } from '@sparkpost/design-tokens-hibana';

export default {
  new: {
    OGFill: '#88E2E7',
    hibanaFill: tokens.color_blue_500,
    OGStroke: '#88E2E7',
    hibanaStroke: tokens.color_blue_500,
    label: 'New',
    description: 'Never engaged, first email in last 7 days',
  },
  uneng: {
    OGFill: '#50CFDA',
    hibanaFill: tokens.color_blue_600,
    OGStroke: '#50CFDA',
    hibanaStroke: tokens.color_blue_600,
    label: 'Never Engaged',
    description: '365+ days since last engagement',
  },
  '365d': {
    OGFill: '#29B9C7',
    hibanaFill: tokens.color_blue_700,
    OGStroke: '#29B9C7',
    hibanaStroke: tokens.color_blue_700,
    label: 'Not Recently Engaged',
    description: '90-365 days since last engagement',
  },
  '90d': {
    OGFill: '#219EA8',
    hibanaFill: tokens.color_blue_800,
    OGStroke: '#219EA8',
    hibanaStroke: tokens.color_blue_800,
    label: 'Semi Recently Engaged',
    description: '14-90 days since last engagement',
  },
  '14d': {
    OGFill: '#1A838B',
    hibanaFill: tokens.color_blue_900,
    OGStroke: '#1A838B',
    hibanaStroke: tokens.color_blue_900,
    label: 'Recently Engaged',
    description: '0-14 days since last engagement',
  },
};
