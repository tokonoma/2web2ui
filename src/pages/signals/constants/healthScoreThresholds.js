import React from 'react';
import { tokens } from '@sparkpost/design-tokens-hibana';
import { CheckCircleOutline, ErrorOutline } from '@sparkpost/matchbox-icons';

const thresholds = {
  danger: {
    color: tokens.color_red_700,
    barColor: tokens.color_red_700,
    icon: ErrorOutline,
    condition: n => n < 55,
    description: (
      <span data-id="health-score-below-55-description">
        A Health score below 55 is considered to be <strong>bad</strong>.
      </span>
    ),
  },
  warning: {
    color: tokens.color_yellow_400,
    barColor: tokens.color_yellow_400,
    icon: ErrorOutline,
    condition: n => n < 80 && n >= 55,
    description: (
      <span data-id="health-score-below-80-description">
        A Health score below 80 <strong>needs improvement</strong>.
      </span>
    ),
  },
  good: {
    color: tokens.color_green_800,
    barColor: tokens.color_green_800,
    icon: CheckCircleOutline,
    condition: n => n >= 80,
    description: (
      <span data-id="health-score-above-80-description">
        A Health score above 80 is considered to be <strong>good</strong>.
      </span>
    ),
  },
};

export default thresholds;

export const HEALTH_SCORE_COLORS_V3 = {
  danger: '#D1E4F4',
  warning: '#91C5FD',
  good: '#4194ED',
};
