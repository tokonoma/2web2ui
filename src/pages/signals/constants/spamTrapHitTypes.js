import { rekey, toCollection } from 'src/helpers/object';
import { tokens } from '@sparkpost/design-tokens-hibana';

// TODO: Remove multiple fill values when OG theme is removed
const spamTrapHitTypes = {
  trap_hits_parked: {
    OGFill: '#50CFDA',
    hibanaFill: tokens.color_blue_900,
    label: 'Parked',
    description: `
      This is an email address from a domain hosted by a parked domain provider. These addresses
      can point to list quality issues but do not affect your reputation.
    `,
  },
  trap_hits_recycled: {
    OGFill: '#29B9C7',
    hibanaFill: tokens.color_blue_700,
    label: 'Recycled',
    description: `
      An email address that once was a real address, but now is a spam trap. These addresses signal
      that you should clean out old unengaged recipients.
    `,
  },
  trap_hits_typo: {
    OGFill: '#1A838B',
    hibanaFill: tokens.color_blue_500,
    label: 'Typo',
    description: `
      This is an email address from a domain that looks like a real mailbox provider, like
      gmal.com. These addresses signal that you should improve your list acquisition practices.
    `,
  },
};

export const spamTrapHitTypesCollection = toCollection(spamTrapHitTypes);
export const spamTrapHitTypesByLabel = rekey(spamTrapHitTypes, 'label');

export default spamTrapHitTypes;
