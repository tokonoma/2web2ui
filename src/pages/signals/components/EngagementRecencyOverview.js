/* eslint-disable max-lines */
import _ from 'lodash';
import React from 'react';
import { Panel } from 'src/components/matchbox';
import SummaryTable, { Column } from 'src/components/summaryTable';
import { setSubaccountQuery } from 'src/helpers/subaccounts';
import { DEFAULT_VIEW } from '../constants/summaryTables';
import FacetDataCell from './dataCells/FacetDataCell';
import NumericDataCell from './dataCells/NumericDataCell';
import PercentDataCell from './dataCells/PercentDataCell';
import SparklineDataCell from './dataCells/SparklineDataCell';
import WoWDataCell from './dataCells/WoWDataCell';
import WoWHeaderCell from './dataCells/WoWHeaderCell';
import Calculation from './viewControls/Calculation';
import styles from './SpamTrapOverview.module.scss';

class EngagementRecencyOverview extends React.Component {
  state = {
    calculation: 'relative',
  };

  componentDidMount() {
    this.resetTable();
  }

  // assumptions, signalOptions and summaryTable should never both change on the same update and
  // resetting signal options will trigger a summary table reset which calls getData
  componentDidUpdate(prevProps) {
    const { signalOptions, summaryTable } = this.props;

    if (prevProps.signalOptions !== signalOptions) {
      this.resetTable();
    }

    if (prevProps.summaryTable !== summaryTable) {
      this.getData();
    }
  }

  resetTable = () => {
    const { defaults = {}, facet, resetSummaryTable, tableName } = this.props;
    let options;

    if (facet.key === 'sid') {
      options = DEFAULT_VIEW;
    }

    resetSummaryTable(tableName, { ...options, ...defaults });
  };

  getData = () => {
    const { getEngagementRecency, signalOptions, summaryTable } = this.props;
    let { subaccount } = signalOptions;
    let order;
    let orderBy;

    if (summaryTable.order) {
      order = summaryTable.order.ascending ? 'asc' : 'desc';
      orderBy = summaryTable.order.dataKey;
    }

    if (subaccount && subaccount.id === undefined) {
      subaccount = undefined; // unset
    }

    getEngagementRecency({
      facet: signalOptions.facet,
      filter: signalOptions.facetSearchTerm,
      from: signalOptions.from,
      limit: summaryTable.perPage,
      offset: (summaryTable.currentPage - 1) * summaryTable.perPage,
      order,
      orderBy,
      relativeRange: signalOptions.relativeRange,
      subaccount,
      to: signalOptions.to,
    });
  };

  handleCalculationChange = calculation => {
    this.setState({ calculation });
  };

  handleClick = facetId => ({ date }) => {
    const { facet, history, signalOptions } = this.props;
    let search;

    if (signalOptions.subaccount) {
      search = setSubaccountQuery(signalOptions.subaccount.id);
    }

    history.push({
      pathname: `/signals/engagement/cohorts/${facet.key}/${facetId}`,
      search,
      state: {
        date,
      },
    });
  };

  render() {
    const {
      data,
      error,
      facet,
      loading,
      signalOptions,
      subaccounts,
      tableName,
      totalCount,
    } = this.props;
    const { calculation } = this.state;
    const isCustomRange = signalOptions.relativeRange === 'custom';

    return (
      <Panel>
        <Panel.Section>
          <div className={styles.Controls}>
            <Calculation initialSelected={calculation} onChange={this.handleCalculationChange} />
          </div>
        </Panel.Section>
        <SummaryTable
          data={data}
          empty={data.length === 0}
          error={error && error.message}
          loading={loading}
          tableName={tableName}
          totalCount={totalCount}
        >
          <Column
            dataKey={facet.key}
            label={facet.label}
            sortable
            width="30%"
            component={props => {
              const id = props[facet.key];

              return (
                <FacetDataCell
                  dimension="engagement/cohorts"
                  facet={facet.key}
                  id={id}
                  name={_.get(_.find(subaccounts, { id }), 'name')}
                  subaccountId={_.get(signalOptions, 'subaccount.id')}
                  truncate
                />
              );
            }}
          />
          <Column
            dataKey="history"
            label="Daily Recently Engaged Recipients"
            width="30%"
            component={({ history, ...data }) => {
              const id = data[facet.key];

              return (
                <SparklineDataCell
                  data={history}
                  dataKey={
                    calculation === 'relative'
                      ? 'relative_engaged_recipients'
                      : 'engaged_recipients'
                  }
                  label="Recently Engaged"
                  onClick={this.handleClick(id)}
                  relative={calculation === 'relative'}
                />
              );
            }}
          />
          {calculation === 'relative' ? (
            <Column
              align="right"
              dataKey="current_relative_engaged_recipients"
              label={isCustomRange ? 'Ratio' : 'Current Ratio'}
              sortable
              width="12.5%"
              component={({ current_relative_engaged_recipients }) => (
                <PercentDataCell value={current_relative_engaged_recipients} />
              )}
            />
          ) : (
            <Column
              align="right"
              dataKey="current_engaged_recipients"
              label={isCustomRange ? 'Count' : 'Current Count'}
              sortable
              width="12.5%"
              component={({ current_engaged_recipients }) => (
                <NumericDataCell value={current_engaged_recipients} />
              )}
            />
          )}
          <Column
            align="right"
            dataKey="WoW"
            label={<WoWHeaderCell />}
            width="12.5%"
            component={({ WoW }) => <WoWDataCell value={WoW} />}
          />
          <Column
            align="right"
            dataKey="c_total"
            label="Current Injections"
            sortable
            width="15%"
            component={({ current_c_total }) => <NumericDataCell value={current_c_total} />}
          />
        </SummaryTable>
      </Panel>
    );
  }
}

export default EngagementRecencyOverview;
