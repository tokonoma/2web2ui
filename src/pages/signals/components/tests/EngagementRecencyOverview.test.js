import React from 'react';
import { shallow } from 'enzyme';
import SummaryTable from 'src/components/summaryTable';
import { DEFAULT_VIEW } from '../../constants/summaryTables';
import EngagementRecencyOverview from '../EngagementRecencyOverview';

describe('EngagementRecencyOverview', () => {
  const subject = (props = {}) =>
    shallow(
      <EngagementRecencyOverview
        data={[
          {
            current_engaged_recipients: 12,
            current_relative_engaged_recipients: 5,
            domain: 'example.com',
            history: [
              { date: '2018-01-13', relative_engaged_recipients: 5, engaged_recipients: 12 },
            ],
            WoW: 0.1,
          },
        ]}
        facet={{
          key: 'domain',
          label: 'Domain',
        }}
        getEngagementRecency={() => {}}
        loading={false}
        resetSummaryTable={() => {}}
        signalOptions={{
          facet: 'domain',
          facetSearchTerm: 'example.com',
          from: '2015-01-01',
          relativeRange: '14days',
          subaccount: {
            id: 123,
          },
          to: '2015-01-05',
        }}
        subaccounts={{
          123: { id: 123, name: 'Test Subaccount' },
        }}
        summaryTable={{
          currentPage: 1,
          order: { ascending: true, dataKey: 'domain' },
          perPage: 10,
        }}
        tableName="Test"
        totalCount={1}
        {...props}
      />,
    );

  it('renders overview panel with controls and table', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('renders empty summary table', () => {
    const wrapper = subject({ data: [] });
    expect(wrapper.find(SummaryTable).prop('empty')).toEqual(true);
  });

  it('renders error message', () => {
    const wrapper = subject({
      data: [],
      error: new Error('Oh no!'),
    });

    expect(wrapper.find(SummaryTable).prop('error')).toEqual('Oh no!');
  });

  it('renders loader', () => {
    const wrapper = subject({
      data: [],
      loading: true,
    });

    expect(wrapper.find(SummaryTable).prop('loading')).toEqual(true);
  });

  it('renders custom date range', () => {
    const wrapper = subject({ signalOptions: { relativeRange: 'custom' } });
    expect(
      wrapper.find('Column[dataKey="current_relative_engaged_recipients"]').prop('label'),
    ).toEqual('Ratio');
  });

  it('handles calculation change', () => {
    const wrapper = subject();
    wrapper.find('Calculation').simulate('change', 'absolute');
    expect(wrapper.state('calculation')).toEqual('absolute');
    expect(wrapper.find('Column[dataKey="current_engaged_recipients"]')).toMatchSnapshot();
  });

  it('renders custom date range with absolute calculation', () => {
    const wrapper = subject({ signalOptions: { relativeRange: 'custom' } });
    wrapper.find('Calculation').simulate('change', 'absolute');
    expect(wrapper.find('Column[dataKey="current_engaged_recipients"]').prop('label')).toEqual(
      'Count',
    );
  });

  it('requests reset on mount', () => {
    const resetSummaryTable = jest.fn();
    subject({ resetSummaryTable });
    expect(resetSummaryTable).toHaveBeenCalledWith('Test', {});
  });

  it('requests reset on mount with default options', () => {
    const resetSummaryTable = jest.fn();
    subject({ resetSummaryTable, defaults: { perPage: 25 } });
    expect(resetSummaryTable).toHaveBeenCalledWith('Test', { perPage: 25 });
  });

  it('requests reset to default view for subaccount view on mount', () => {
    const resetSummaryTable = jest.fn();
    subject({ facet: { key: 'sid', label: 'Subaccounts' }, resetSummaryTable });
    expect(resetSummaryTable).toHaveBeenCalledWith('Test', DEFAULT_VIEW);
  });

  it('requests table reset on signal options update', () => {
    const resetSummaryTable = jest.fn();
    const wrapper = subject();
    wrapper.setProps({ resetSummaryTable, signalOptions: {} });

    expect(resetSummaryTable).toHaveBeenCalledWith('Test', {});
  });

  it('requests data on summary table update', () => {
    const getEngagementRecency = jest.fn();
    const summaryTable = {
      currentPage: 2,
      perPage: 10,
    };
    const wrapper = subject();

    wrapper.setProps({ getEngagementRecency, summaryTable });

    expect(getEngagementRecency).toHaveBeenCalledWith({
      facet: 'domain',
      filter: 'example.com',
      from: '2015-01-01',
      limit: 10,
      offset: 10,
      order: undefined,
      orderBy: undefined,
      relativeRange: '14days',
      subaccount: {
        id: 123,
      },
      to: '2015-01-05',
    });
  });

  it('requests ordered data on summary table update', () => {
    const getEngagementRecency = jest.fn();
    const summaryTable = {
      currentPage: 1,
      order: { ascending: true, dataKey: 'domain' },
      perPage: 10,
    };
    const wrapper = subject();

    wrapper.setProps({ getEngagementRecency, summaryTable });

    expect(getEngagementRecency).toHaveBeenCalledWith(
      expect.objectContaining({
        order: 'asc',
        orderBy: 'domain',
      }),
    );
  });

  it('requests data without subaccount data on summary table update', () => {
    const getEngagementRecency = jest.fn();
    const signalOptions = {
      facet: 'domain',
      facetSearchTerm: 'example.com',
      relativeRange: '14days',
      subaccount: {
        id: undefined,
        name: 'Master & All Subaccounts',
      },
    };
    const summaryTable = {
      currentPage: 2,
      perPage: 10,
    };
    const wrapper = subject();

    wrapper.setProps({ getEngagementRecency, signalOptions, summaryTable });

    expect(getEngagementRecency).toHaveBeenCalledWith(
      expect.objectContaining({
        subaccount: undefined,
      }),
    );
  });

  describe('history component', () => {
    const factory = ({ calculation, history, ...props }) => {
      const wrapper = subject({
        history,
        metaData: { currentMax: 200, currentRelativeMax: 40 },
      });
      wrapper.setState({ calculation });
      const Column = wrapper.find('Column[dataKey="history"]').prop('component');

      return shallow(<Column {...props} domain="example.com" />);
    };

    it('renders absolute sparkline', () => {
      const wrapper = factory({ chartType: 'line' });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders relative sparkline', () => {
      const wrapper = factory({ calculation: 'relative', chartType: 'line' });
      expect(wrapper).toMatchSnapshot();
    });

    it('redirects to details page when bar is clicked', () => {
      const historyPush = jest.fn();
      const wrapper = factory({ chartType: 'bar', history: { push: historyPush } });
      wrapper.simulate('click', { date: '2018-01-13' });

      expect(historyPush).toHaveBeenCalledWith({
        pathname: '/signals/engagement/cohorts/domain/example.com',
        search: '?subaccount=123',
        state: { date: '2018-01-13' },
      });
    });

    it('redirects to details page when dot is clicked', () => {
      const historyPush = jest.fn();
      const wrapper = factory({ chartType: 'line', history: { push: historyPush } });
      wrapper.simulate('click', { date: '2018-01-13' });

      expect(historyPush).toHaveBeenCalledWith({
        pathname: '/signals/engagement/cohorts/domain/example.com',
        search: '?subaccount=123',
        state: { date: '2018-01-13' },
      });
    });
  });

  describe('current column component', () => {
    const factory = ({ calculation, ...props }) => {
      const wrapper = subject();
      wrapper.setState({ calculation });
      const dataKey =
        calculation === 'relative'
          ? 'current_relative_engaged_recipients'
          : 'current_engaged_recipients';
      const Column = wrapper.find(`Column[dataKey="${dataKey}"]`).prop('component');

      return shallow(<Column {...props} />);
    };

    it('renders current count', () => {
      const wrapper = factory({ calculation: 'absolute', current_engaged_recipients: 123 });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders current rate', () => {
      const wrapper = factory({
        calculation: 'relative',
        current_relative_engaged_recipients: 234,
      });
      expect(wrapper).toMatchSnapshot();
    });
  });
});
