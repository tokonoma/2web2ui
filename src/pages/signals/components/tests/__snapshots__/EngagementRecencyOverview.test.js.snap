// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagementRecencyOverview current column component renders current count 1`] = `
<NumericDataCell
  value={123}
/>
`;

exports[`EngagementRecencyOverview current column component renders current rate 1`] = `
<PercentDataCell
  value={234}
/>
`;

exports[`EngagementRecencyOverview handles calculation change 1`] = `
<Column
  align="right"
  component={[Function]}
  dataKey="current_engaged_recipients"
  label="Current Count"
  sortable={true}
  width="12.5%"
/>
`;

exports[`EngagementRecencyOverview history component renders absolute sparkline 1`] = `
<SparklineDataCell
  dataKey="engaged_recipients"
  label="Recently Engaged"
  onClick={[Function]}
  relative={false}
/>
`;

exports[`EngagementRecencyOverview history component renders relative sparkline 1`] = `
<SparklineDataCell
  dataKey="relative_engaged_recipients"
  label="Recently Engaged"
  onClick={[Function]}
  relative={true}
/>
`;

exports[`EngagementRecencyOverview renders overview panel with controls and table 1`] = `
<Panel>
  <Panel.Section>
    <div
      className="Controls"
    >
      <Calculation
        initialSelected="relative"
        onChange={[Function]}
      />
    </div>
  </Panel.Section>
  <Connect(SummaryTable)
    data={
      Array [
        Object {
          "WoW": 0.1,
          "current_engaged_recipients": 12,
          "current_relative_engaged_recipients": 5,
          "domain": "example.com",
          "history": Array [
            Object {
              "date": "2018-01-13",
              "engaged_recipients": 12,
              "relative_engaged_recipients": 5,
            },
          ],
        },
      ]
    }
    empty={false}
    loading={false}
    tableName="Test"
    totalCount={1}
  >
    <Column
      component={[Function]}
      dataKey="domain"
      label="Domain"
      sortable={true}
      width="30%"
    />
    <Column
      component={[Function]}
      dataKey="history"
      label="Daily Recently Engaged Recipients"
      width="30%"
    />
    <Column
      align="right"
      component={[Function]}
      dataKey="current_relative_engaged_recipients"
      label="Current Ratio"
      sortable={true}
      width="12.5%"
    />
    <Column
      align="right"
      component={[Function]}
      dataKey="WoW"
      label={<WoWHeaderCell />}
      width="12.5%"
    />
    <Column
      align="right"
      component={[Function]}
      dataKey="c_total"
      label="Current Injections"
      sortable={true}
      width="15%"
    />
  </Connect(SummaryTable)>
</Panel>
`;
